packages/block-library/src/embed/edit.js

```diff
@@ -121,123 +65,137 @@ export function getEmbedEditComponent( title, icon, responsive = true ) {
 					allowResponsive
 				),
 			};
-		}
-
-		/***
-		 * Sets block attributes based on the current attributes and preview data.
-		 */
-		setMergedAttributes() {
-			const { setAttributes } = this.props;
-			setAttributes( this.getMergedAttributes() );
-		}
+		};
 
-		switchBackToURLInput() {
-			this.setState( { editingURL: true } );
-		}
+		const handleIncomingPreview = () => {
+			setAttributes( getMergedAttributes() );
+			if ( onReplace ) {
+				const upgradedBlock = createUpgradedEmbedBlock(
+					props,
+					getMergedAttributes()
+				);
 
-		getResponsiveHelp( checked ) {
-			return checked
-				? __(
-						'This embed will preserve its aspect ratio when the browser is resized.'
-				  )
-				: __(
-						'This embed may not preserve its aspect ratio when the browser is resized.'
-				  );
-		}
+				if ( upgradedBlock ) {
+					onReplace( upgradedBlock );
+				}
+			}
+		};
 
-		toggleResponsive() {
-			const { allowResponsive, className } = this.props.attributes;
-			const { html } = this.props.preview;
+		const toggleResponsive = () => {
+			const { allowResponsive, className } = attributes;
+			const { html } = preview;
 			const newAllowResponsive = ! allowResponsive;
 
-			this.props.setAttributes( {
+			setAttributes( {
 				allowResponsive: newAllowResponsive,
 				className: getClassNames(
 					html,
 					className,
 					responsive && newAllowResponsive
 				),
 			} );
-		}
+		};
 
-		render() {
-			const { url, editingURL } = this.state;
-			const {
-				fetching,
-				setAttributes,
-				isSelected,
-				preview,
-				cannotEmbed,
-				themeSupportsResponsive,
-				tryAgain,
-			} = this.props;
+		useEffect( () => {
+			// If we can embed the url, bail early.
+			if ( ! cannotEmbed ) {
+				return;
+			}
 
+			// At this stage, we either have a new preview or a new URL, but we can't embed it.
+			// If we are already fetching the preview, bail early.
 			if ( fetching ) {
-				return <EmbedLoading />;
+				return;
 			}
 
-			// translators: %s: type of embed e.g: "YouTube", "Twitter", etc. "Embed" is used when no specific type exists
-			const label = sprintf( __( '%s URL' ), title );
-
-			// No preview, or we can't embed the current URL, or we've clicked the edit button.
-			if ( ! preview || cannotEmbed || editingURL ) {
-				return (
-					<EmbedPlaceholder
-						icon={ icon }
-						label={ label }
-						onSubmit={ this.setUrl }
-						value={ url }
-						cannotEmbed={ cannotEmbed }
-						onChange={ ( event ) =>
-							this.setState( { url: event.target.value } )
-						}
-						fallback={ () => fallback( url, this.props.onReplace ) }
-						tryAgain={ tryAgain }
-					/>
-				);
+			// At this stage, we're not fetching the preview, so we know it can't be embedded, so try
+			// removing any trailing slash, and resubmit.
+			const newURL = attributes.url.replace( /\/$/, '' );
+			setURL( newURL );
+			setIsEditingURL( false );
+			setAttributes( { url: newURL } );
+		}, [ cannotEmbed, fetching, attributes ] );
```

Comment:
You forgot to include `preview.html` in the dependency array. Also, this could be a bit more optimized:
```suggestion
		}, [ attributes.allowResponsive, attributes.className, cannotEmbed, fetching, preview.html ] );
```